src/iidm/IdBasedBusRef.cpp

```diff
@@ -0,0 +1,73 @@
+/**
+ * Copyright (c) 2021, RTE (http://www.rte-france.com)
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/.
+ */
+
+#include <powsybl/iidm/IdBasedBusRef.hpp>
+
+#include <powsybl/iidm/Bus.hpp>
+#include <powsybl/iidm/Enum.hpp>
+#include <powsybl/iidm/Identifiable.hpp>
+#include <powsybl/iidm/Network.hpp>
+#include <powsybl/stdcxx/instanceof.hpp>
+
+namespace powsybl {
+
+namespace iidm {
+
+IdBasedBusRef::IdBasedBusRef(const std::string& id) :
+    m_id(id) {
+}
+
+IdBasedBusRef::IdBasedBusRef(const std::string& id, const Branch::Side& side) :
+    m_id(id), m_side(side) {
+}
+
+stdcxx::CReference<Bus> IdBasedBusRef::resolveByLevel(const Network& network, const TopologyLevel& level) {
+    if (!m_side) {
+        const auto& identifiable = network.find(m_id);
+        if (!identifiable) {
+            return stdcxx::cref<Bus>();
+        }
+        if (stdcxx::isInstanceOf<Bus>(identifiable.get())) {
+            const auto& bus = dynamic_cast<const Bus&>(identifiable.get());
+            if (level == TopologyLevel::BUS_BRANCH) {
+                for (const auto& terminal : bus.getConnectedTerminals()) {
+                    if (terminal.getBusView().getBus()) {
+                        return terminal.getBusView().getBus();
+                    }
+                    return stdcxx::cref<Bus>();
+                }
+            } else {
+                return stdcxx::cref(bus);
+            }
+        } else if (stdcxx::isInstanceOf<Injection>(identifiable.get())) {
+            const auto& injection = dynamic_cast<const Injection&>(identifiable.get());
+            return chooseBusByLevel(injection.getTerminal(), level);
+        } else {
+            throw PowsyblException(stdcxx::format("%1% is not a bus or injection.", m_id));
+        }
+    }
+    const auto& branch = network.find<Branch>(m_id);
+    if (!branch) {
```

Comment:
It looks like there is a bug in java if `Branch branch = network.getBranch(id);` return `null` (which may occurs if id does not exist in the network)

Reply:
Indeed. I fixed the issue you mentioned in this PR: https://github.com/powsybl/powsybl-core/pull/1821

Reply:
Allright, I backported it.